gst/nnstreamer/tensor_filter/tensor_filter_common.c

```diff
@@ -2786,3 +2801,126 @@ gst_tensor_filter_check_hw_availability (const gchar * name, const accl_hw hw)
+ * @return 0 if the new key and value are inserted. Others are failed to insert it.
+ */
+int
+nnstreamer_filter_shared_table_insert (char *key, void *value, void *instance)
+{
+  GstTensorFilterSharedModelRepresenatation *model_rep;
+  gboolean is_existed = FALSE;
+
+  if (!shared_model_table) {
+    ml_loge ("The shared model representation is not supported properly!");
+    return FALSE;
+  }
+  if (!key) {
+    ml_loge ("The key should NOT be NULL!");
+    return FALSE;
+  }
+  if (!instance) {
+    ml_loge ("The instance should NOT be NULL!");
+    return FALSE;
+  }
+
+  model_rep = g_hash_table_lookup (shared_model_table, key);
+  is_existed = model_rep != NULL;
+
+  if (is_existed) {
+    model_rep->referred_list =
+        g_list_append (model_rep->referred_list, instance);
+    return TRUE;
+  } else {
+    if (!value) {
+      ml_loge ("The value should NOT be NULL!");
+      return FALSE;
+    }
+
+    model_rep = (GstTensorFilterSharedModelRepresenatation *)
+        g_malloc0 (sizeof (GstTensorFilterSharedModelRepresenatation));
+    model_rep->shared_interpreter = value;
+    model_rep->referred_list =
+        g_list_append (model_rep->referred_list, instance);
+
+    return g_hash_table_insert (shared_model_table, g_strdup (key),
+        (gpointer) model_rep);
+  }
+}
+
+/* extern functions for shared model representation table */
+/**
+ * @brief Remove the instance registered at the referred list of shared model table.
+ *        If referred list is empty, remove the shared representation from the table.
+ * @param[in] key The key to find the matched list of hash table.
+ * @param[in] instance The instance that should be removed from the referred list.
+ * @param[in] free_callback The callback function to delete internal objects of instance.
+ * @return TRUE if the new key and value are removed. FALSE if failed to remove it.
+ */
+int
+nnstreamer_filter_shared_table_remove (const char *key, void *instance,
+    void (*free_callback) (void *))
+{
+  GstTensorFilterSharedModelRepresenatation *value;
+  int ret = TRUE;
+
+  /* search the table with key */
+  if (!shared_model_table) {
+    ml_loge ("The shared model representation is not supported properly!");
+    return FALSE;
+  }
+
+  value = g_hash_table_lookup (shared_model_table, key);
+  if (!value) {
+    ml_loge ("There is no value of the key: %s", key);
+    return FALSE;
+  }
+
+  /* remove instance from the list */
+  g_mutex_lock (&mutex);
+  value->referred_list = g_list_remove (value->referred_list, instance);
+  g_mutex_unlock (&mutex);
+  ml_logd ("The referred instance of sharing key: %s has been removed!", key);
+
+  /* remove key from table if list is empty */
+  if (g_list_length (value->referred_list) == 0) {
+    g_list_free (value->referred_list);
+    g_mutex_lock (&mutex);
+    ret = g_hash_table_remove (shared_model_table, key);
+    if (free_callback)
+      free_callback (instance);
+    g_mutex_unlock (&mutex);
```

Comment:
The protected resource, "shared_model_table", is not protected properly. Look at line 2878.